README.md
Single-beat
---------
Single-beat is a nice little application that ensures only one instance of your process runs across your servers.

Such as celerybeat (or some kind of daily mail sender, orphan file cleaner etc...) needs to be running only on one server,
but if that server gets down, well, you go and start it at another server etc. 

As we all hate manually doing things, single-beat automates this process.


How
---------

We use redis as a lock server, and wrap your process with single-beat, in two servers,

```bash
single-beat celery beat
```

on the second server

```bash
single-beat celery beat
```

on the third server

```bash
single-beat celery beat
```

The second process will just wait until the first one dies etc.

Installation
------------

```bash
sudo pip install single-beat
```

Configuration
-------------

You can configure single-beat with environment variables, like

```bash
SINGLE_BEAT_REDIS_SERVER='redis://redis-host:6379/1' single-beat celery beat
```

- SINGLE_BEAT_REDIS_SERVER

    you can give redis host url, we pass it to from_url of [redis-py](http://redis-py.readthedocs.org/en/latest/#redis.StrictRedis.from_url)

- SINGLE_BEAT_REDIS_PASSWORD

    for usage in sentinel scenarios (since they ignore SINGLE_BEAT_REDIS_SERVER)

- SINGLE_BEAT_REDIS_SENTINEL

    use redis sentinel to select the redis host to use, sentinels are defined as colon-separated list of hostname and port pairs, e.g. `192.168.1.10:26379;192.168.1.11:26379;192.168.1.12:26379`

- SINGLE_BEAT_REDIS_SENTINEL_MASTER (default `mymaster`)
- SINGLE_BEAT_REDIS_SENTINEL_DB (default 0)
- SINGLE_BEAT_REDIS_SENTINEL_PASSWORD
- SINGLE_BEAT_IDENTIFIER

    the default is we use your process name as the identifier, like

    ```bash
    single-beat celery beat
    ```

    all processes checks a key named, SINGLE_BEAT_celery but in some cases you might need to give another identifier, eg. your project name etc.

    ```bash
    SINGLE_BEAT_IDENTIFIER='celery-beat' single-beat celery beat
    ```

- SINGLE_BEAT_LOCK_TIME (default 5 seconds)
- SINGLE_BEAT_INITIAL_LOCK_TIME (default 2 * SINGLE_BEAT_LOCK_TIME seconds)
- SINGLE_BEAT_HEARTBEAT_INTERVAL (default 1 second)

    when starting your process, we set a key with 10 second expiration (INITIAL_LOCK_TIME) in redis server, 
    other single-beat processes checks if that key exists - if it exists they won't spawn children. 
    
    We continue to update that key every 1 second (HEARTBEAT_INTERVAL) setting it with a ttl of 5 seconds (LOCK_TIME)

    This should work, but you might want to give more relaxed intervals, like:

    ```bash
    SINGLE_BEAT_LOCK_TIME=300 SINGLE_BEAT_HEARTBEAT_INTERVAL=60 single-beat celery beat
    ```

- SINGLE_BEAT_HOST_IDENTIFIER (default socket.gethostname)

    we set the name of the host and the process id as lock keys value, so you can check where your process lives.

    ```bash
    SINGLE_BEAT_IDENTIFIER='celery-beat' single-beat celery beat
    ```

    ```bash
    (env)$ redis-cli
    redis 127.0.0.1:6379> keys *
    1) "_kombu.binding.celeryev"
    2) "celery"
    3) "_kombu.binding.celery"
    4) "SINGLE_BEAT_celery-beat"
    redis 127.0.0.1:6379> get SINGLE_BEAT_celery-beat
    "0:aybarss-MacBook-Air.local:43213"
    redis 127.0.0.1:6379>
    ```

    ```bash
    SINGLE_BEAT_HOST_IDENTIFIER='192.168.1.1' SINGLE_BEAT_IDENTIFIER='celery-beat' single-beat celery beat
    ```

    ```bash
    (env)$ redis-cli
    redis 127.0.0.1:6379> keys *
    1) "SINGLE_BEAT_celery-beat"
    redis 127.0.0.1:6379> get SINGLE_BEAT_celery-beat
    "0:192.168.1.1:43597"
    ```

- SINGLE_BEAT_LOG_LEVEL (default warn)

    change log level to debug if you want to see the heartbeat messages.

- SINGLE_BEAT_WAIT_MODE (default heartbeat)
- SINGLE_BEAT_WAIT_BEFORE_DIE (default 60 seconds)
    
    singlebeat has two different modes:
        - heartbeat (default)
        - supervised
    
    In heartbeat mode, single-beat is responsible for everything, spawning a process checking its status, publishing etc.
    In supervised mode, single-beat starts, checks if the child is running somewhere and waits for a while and then exits. So supervisord - or another scheduler picks up and restarts single-beat.

    on first server

    ```bash
    SINGLE_BEAT_LOG_LEVEL=debug SINGLE_BEAT_WAIT_MODE=supervised SINGLE_BEAT_WAIT_BEFORE_DIE=10 SINGLE_BEAT_IDENTIFIER='celery-beat' single-beat celery beat -A example.tasks
    DEBUG:singlebeat.beat:timer called 0.100841999054 state=WAITING
    [2014-05-05 16:28:24,099: INFO/MainProcess] beat: Starting...
    DEBUG:singlebeat.beat:timer called 0.999553918839 state=RUNNING
    DEBUG:singlebeat.beat:timer called 1.00173187256 state=RUNNING
    DEBUG:singlebeat.beat:timer called 1.00134801865 state=RUNNING
    ```

    this will heartbeat every second, on your second server

    ```bash
    SINGLE_BEAT_LOG_LEVEL=debug SINGLE_BEAT_WAIT_MODE=supervised SINGLE_BEAT_WAIT_BEFORE_DIE=10 SINGLE_BEAT_IDENTIFIER='celery-beat' single-beat celery beat -A example.tasks
    DEBUG:singlebeat.beat:timer called 0.101243019104 state=WAITING
    DEBUG:root:already running, will exit after 60 seconds
    ```

    so if you do this in your supervisor.conf

    ```bash
    [program:celerybeat]
    environment=SINGLE_BEAT_IDENTIFIER="celery-beat",SINGLE_BEAT_REDIS_SERVER="redis://localhost:6379/0",SINGLE_BEAT_WAIT_MODE="supervised", SINGLE_BEAT_WAIT_BEFORE_DIE=10
    command=single-beat celery beat -A example.tasks
    numprocs=1
    stdout_logfile=./logs/celerybeat.log
    stderr_logfile=./logs/celerybeat.err
    autostart=true
    autorestart=true
    startsecs=10
    ```

    it will try to spawn celerybeat every 60 seconds.

Cli
-------------
Single-beat also has a simple cli, that gives info about where your process is living - also can pause single-beat, restart your process etc.

"info" will show where the process is running, the first node identifier is the ip address connecting to redis - by default.

```
(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95779 | WAITING |
127.0.0.1:95776 | RUNNING | pid: 95778
127.0.0.1:95784 | WAITING |
```


"stop", will stop your child process, so any node will pick it up again

```
(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli stop
127.0.0.1:95776 | PAUSED | killed
127.0.0.1:95779 | WAITING |
127.0.0.1:95784 | WAITING |

(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95776 | WAITING |
127.0.0.1:95779 | WAITING |
127.0.0.1:95784 | RUNNING | pid: 95877
```

"restart" will restart the child process in the active node.

```
(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95776 | WAITING |
127.0.0.1:95779 | WAITING |
127.0.0.1:95784 | RUNNING | pid: 95877

(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli restart
127.0.0.1:95776 | WAITING |
127.0.0.1:95779 | WAITING |
127.0.0.1:95784 | RESTARTING | killed

(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95776 | WAITING |
127.0.0.1:95779 | WAITING |
127.0.0.1:95784 | RUNNING | pid: 95905
```


"pause" will kill the child, and put all single-beat nodes in pause state. This is useful for when deploying, to ensure that no "old version of the code"
is running while the deploy process is in place. after the deploy you can "resume" so any node will pick the child.

```
(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95776 | WAITING |
127.0.0.1:95779 | WAITING |
127.0.0.1:95784 | RUNNING | pid: 95905

(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli pause
127.0.0.1:95776 | PAUSED |
127.0.0.1:95779 | PAUSED |
127.0.0.1:95784 | PAUSED | killed

(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95776 | PAUSED |
127.0.0.1:95779 | PAUSED |
127.0.0.1:95784 | PAUSED |
```

"resume" will put single-beat nodes in waiting state - so any node will pick up the child

```
(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95776 | PAUSED |
127.0.0.1:95779 | PAUSED |
127.0.0.1:95784 | PAUSED |

(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli resume
127.0.0.1:95776 | WAITING |
127.0.0.1:95784 | WAITING |
127.0.0.1:95779 | WAITING |

(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
127.0.0.1:95776 | WAITING |
127.0.0.1:95784 | WAITING |
127.0.0.1:95779 | RUNNING | pid: 96025
```


"quit" will terminate all child processes and then the parent process itself. So there will be no live single-beat nodes. Its useful to
have some sort of hand-brake - also might be useful when you have blue/green, or canary style deployments.

```
(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli quit
127.0.0.1:95784 | RUNNING |

(venv3) $
(venv3) $ SINGLE_BEAT_IDENTIFIER=echo SINGLE_BEAT_LOG_LEVEL=critical SINGLE_BEAT_REDIS_SERVER=127.0.0.1 single-beat-cli info
```


Usage Patterns
--------------

You can see an example usage with supervisor at example/celerybeat.conf

Why
--------

There are some other solutions but either they are either complicated, or you need to modify the process. And I couldn't find a simpler solution for this https://github.com/celery/celery/issues/251 without modifying or adding locks to my tasks.

You can also check uWsgi's [Legion Support](http://uwsgi-docs.readthedocs.org/en/latest/AttachingDaemons.html#legion-support) which can do the same thing.

Credits
----------
 * [ybrs](https://github.com/ybrs)
 * [edmund-wagner](https://github.com/edmund-wagner)
 * [lowks](https://github.com/lowks)
 * [rangermeier](https://github.com/rangermeier)
 * [joekohlsdorf](https://github.com/joekohlsdorf)
 * [robvdl](https://github.com/robvdl)
 * [neuroforge](https://github.com/neuroforgede/single-beat)
 * [chripede](https://github.com/chripede/)
 

ChangeLog
Version 0.6.3
==============
 - Release issue, re-releasing 0.6.2

Version 0.6.2
==============
 - Fixes asyncio/python 3.11 issues. 

Version 0.6.1
==============
 - Fixes unicode decode errors - if split multibyte char at 100th character. see https://github.com/ybrs/single-beat/pull/48

Version 0.6.0
==============
 - Use asyncio

Version 0.5.1
==============
 - guard stdout/stderr forwarding against crashes

Version 0.5.0
==============
 - Enforce tornado >= 6.0
 - Switch redis library to aioredis
 - remove tornadis dependency
 - change async logic to use standard python async/await

Version 0.4.2
==============
 - Fixes some small bug, typos in documentation.

Version 0.4.1
==============
 - Added a simple cli to remotely control single-beat instances

Version 0.3.2
==============
 - Try to reacquire Redis lock when it expired

Version 0.3.1
==============
 - Fixes python 3 compatibility issue

Version 0.3.0
==============
 - Add fencing
 
Version 0.2.1
==========
 - Support Redis Sentinel

Version 0.2.0
==============
 - Moved to tornado's ioloop, pyuv was hard to install, had breaking changes etc.
 - Code cleanup/refactor
 - Simple unit test
 - Started putting a changelog


test/echo.py
import sys
import time
import logging

logging.basicConfig(level=logging.DEBUG)

while True:
    logging.info(time.time())
    try:
        time.sleep(1)
    except KeyboardInterrupt:
        break

print("exiting from echo !")
sys.exit(3)


test/check_output.py
files = ["first_single_beat.log", "second_single_beat.log", "third_single_beat.log"]

hello_cnt = 0
for fname in files:
    f = open(fname)
    if "hello - " in f.read():
        hello_cnt += 1

assert hello_cnt == 1


test/test_process.py
import subprocess
import unittest
import time
import psutil
import os

import signal


class TestStringMethods(unittest.TestCase):

    def setup_processes(self, cmd="single-beat python echo.py", cnt=3):
        self.processes = [
            subprocess.Popen(cmd, shell=True)
            for i in range(0, cnt)
        ]

    def get_all_children(self):
        children = []
        for p in self.processes:
            try:
                parent = psutil.Process(p.pid)
            except psutil.NoSuchProcess:
                continue
            children += parent.children(recursive=True)
        return children

    def get_active_subprocess(self):
        for p in self.processes:
            try:
                parent = psutil.Process(p.pid)
            except psutil.NoSuchProcess:
                continue
            if parent.children(recursive=True):
                return p

    def wait_and_try_again(self, fn, max_seconds=5):
        cnt = 0
        while cnt < max_seconds:
            result = fn()
            if result:
                return result
            time.sleep(1)
            cnt += 1
        raise Exception(f"Couldn't complete {fn} in a timely fashion")

    def _test_single_instance(self):
        """
        only one instance of the child should be running
        """
        self.setup_processes()
        # check we only one echo.py running
        children = self.wait_and_try_again(self.get_all_children, 5)
        self.assertEqual(len(children), 1, "we have more than one child")
        # send kill to the child, see if someone else spawns
        os.kill(children[0].pid, signal.SIGTERM)
        new_children = self.wait_and_try_again(self.get_all_children, 30)
        self.assertGreater(len(children), 0, "couldn't spawn new child in 30 seconds")
        self.assertEqual(len(children), 1, "we have more than one child")
        self.assertNotEqual(
            children[0].pid,
            new_children[0].pid,
            "couldn't kill child - pid should've changed",
        )

    def test_exit_code(self):
        self.setup_processes("single-beat python echo-exit-code.py", cnt=1)
        # check we only one echo.py running
        children = self.wait_and_try_again(self.get_all_children, 5)
        self.assertEqual(len(children), 1, "we have more than one child")
        sp = self.get_active_subprocess()
        exit_code = self.wait_and_try_again(lambda: sp.poll(), 5)
        self.assertEqual(exit_code, 3)

    def tearDown(self):
        for p in self.processes:
            try:
                p.terminate()
                p.wait(10)
            except Exception as e:
                import traceback

                traceback.print_exc()


if __name__ == "__main__":
    unittest.main()


singlebeat/cli.py
from .beat import config
import json
import uuid
import time
import threading
import click

reply_channel = 'SINGLE_BEAT_REPLY_{}'.format(uuid.uuid4())


class Commander(object):

    def cmd_info(self):
        return 'info', []

    def cmd_quit(self):
        """\
        kills the child - if running - and then single-beat itself exits
        useful to terminate all single-beat instances in one go
        :return:
        """
        return 'quit', []

    def cmd_pause(self):
        """\
        it will kill the child and pause all nodes
        """
        return 'pause', []

    def cmd_resume(self):
        """\
        will resume all nodes - set them to waiting state
        """
        return 'resume', []

    def cmd_restart(self):
        """\
        it will restart the child process - in the same node
        useful for when deploying new code
        """
        return 'restart', []

    def cmd_stop(self):
        """\
        it will stop the child process, then any single-beat node will pick it up
        and restart
        """
        return 'stop', []


def submit_to_replies():
    rds = config.get_redis()
    cnt = rds.pubsub_numsub('SB_{}'.format(config.IDENTIFIER))
    cnt = cnt[0][1]
    p = rds.pubsub()
    p.subscribe(reply_channel)
    while cnt > 0:
        message = p.get_message(timeout=5)
        if not message:
            return
        if message['type'] == 'subscribe':
            continue

        data = json.loads(message['data'])
        print("{} | {} | {}".format(data['identifier'], data['state'], data.get('info', '') ))
        cnt = cnt - 1

help_text = """
info: 
    Show information about which nodes are holding the running application

quit:
    kills the child - if running - and then single-beat itself exits
useful to terminate all single-beat instances in one go

pause:
    it will kill the child and pause all nodes. Might be useful to deploy new code.

resume:
    will start all nodes, and put them in waiting state - one of them will pick the application

restart:
    it will restart the child process - in the same node useful for when deploying new code
    
stop:
    it will kill the child process so any node will pick it up and start.
"""

@click.command()
@click.argument('cmd', required=False)
def main(cmd='info'):
    commander = Commander()
    fn = getattr(commander, 'cmd_{}'.format(cmd), None)
    if not fn:
        print('unknown command, available commands are')
        print(help_text)
        return
    thread = threading.Thread(target=submit_to_replies)
    thread.start()
    time.sleep(0.100)
    #
    cmd_name, args = fn()
    cmd = json.dumps({
        'cmd': cmd_name,
        'args': args,
        'reply_channel': reply_channel
    })
    rds = config.get_redis()
    rds.publish('SB_{}'.format(config.IDENTIFIER), cmd)


if __name__ == '__main__':
    main()



singlebeat/config.py
import os
import socket

def noop(i):
    return i

def env(identifier, default, type=noop):
    return type(os.getenv('SINGLE_BEAT_%s' % identifier, default))


class Config(object):
    REDIS_SERVER = env('REDIS_SERVER', 'redis://localhost:6379')
    IDENTIFIER = env('IDENTIFIER', None)
    LOCK_TIME = env('LOCK_TIME', 5, int)
    INITIAL_LOCK_TIME = env('INITIAL_LOCK_TIME', LOCK_TIME * 2, int)
    HEARTBEAT_INTERVAL = env('HEARTBEAT_INTERVAL', 1, int)
    HOST_IDENTIFIER = env('HOST_IDENTIFIER', socket.gethostname())
    LOG_LEVEL = env('LOG_LEVEL', 'warn')
    # wait_mode can be, supervisored or heartbeat
    WAIT_MODE = env('WAIT_MODE', 'heartbeat')
    WAIT_BEFORE_DIE = env('WAIT_BEFORE_DIE', 60, int)

    def check(self, cond, message):
        if not cond:
            raise Exception(message)

    def checks(self):
        self.check(self.LOCK_TIME < self.INITIAL_LOCK_TIME, "inital lock time must be greater than lock time")
        self.check(self.HEARTBEAT_INTERVAL < (self.LOCK_TIME / 2.0), "SINGLE_BEAT_HEARTBEAT_INTERVAL must be smaller than SINGLE_BEAT_LOCK_TIME / 2")
        self.check(self.WAIT_MODE in ('supervised', 'heartbeat'), 'undefined wait mode')


singlebeat/beat.py
import codecs
import functools
import json
import os
import sys
import time
import socket
import redis
import redis.sentinel
import logging
import signal
import asyncio
from redis import asyncio as aioredis


def noop(i):
    return i


def env(identifier, default, type=noop):
    return type(os.getenv("SINGLE_BEAT_%s" % identifier, default))


class Config(object):
    REDIS_SERVER = env('REDIS_SERVER', 'redis://localhost:6379')
    REDIS_PASSWORD = env('REDIS_PASSWORD', None)
    REDIS_SENTINEL = env('REDIS_SENTINEL', None)
    REDIS_SENTINEL_MASTER = env('REDIS_SENTINEL_MASTER', 'mymaster')
    REDIS_SENTINEL_DB = env('REDIS_SENTINEL_DB', 0)
    REDIS_SENTINEL_PASSWORD = env('REDIS_SENTINEL_PASSWORD', None)
    IDENTIFIER = env('IDENTIFIER', None)
    LOCK_TIME = env('LOCK_TIME', 5, int)
    INITIAL_LOCK_TIME = env('INITIAL_LOCK_TIME', LOCK_TIME * 2, int)
    HEARTBEAT_INTERVAL = env('HEARTBEAT_INTERVAL', 1, int)
    HOST_IDENTIFIER = env('HOST_IDENTIFIER', socket.gethostname())
    LOG_LEVEL = env('LOG_LEVEL', 'warn')
    # wait_mode can be, supervisord or heartbeat
    WAIT_MODE = env("WAIT_MODE", "heartbeat")
    WAIT_BEFORE_DIE = env("WAIT_BEFORE_DIE", 60, int)
    _host_identifier = None

    def check(self, cond, message):
        if not cond:
            raise Exception(message)

    def checks(self):
        self.check(
            self.LOCK_TIME < self.INITIAL_LOCK_TIME,
            "initial lock time must be greater than lock time",
        )
        self.check(
            self.HEARTBEAT_INTERVAL < (self.LOCK_TIME / 2.0),
            "SINGLE_BEAT_HEARTBEAT_INTERVAL must be smaller than SINGLE_BEAT_LOCK_TIME / 2",
        )
        self.check(self.WAIT_MODE in ("supervised", "heartbeat"), "undefined wait mode")
        if self.REDIS_SENTINEL:
            master = self._sentinel.discover_master(self.REDIS_SENTINEL_MASTER)
        else:
            self._redis.ping()

    def get_redis(self):
        if self.REDIS_SENTINEL:
            return self._sentinel.master_for(self.REDIS_SENTINEL_MASTER,
                                             password=self.REDIS_PASSWORD,
                                             redis_class=redis.Redis)
        return self._redis

    def rewrite_redis_url(self):
        """\
        if REDIS_SERVER is just an ip address, then we try to translate it to
        redis_url, redis://REDIS_SERVER so that it doesn't try to connect to
        localhost while you try to connect to another server
        :return:
        """
        if (
            self.REDIS_SERVER.startswith("unix://")
            or self.REDIS_SERVER.startswith("redis://")
            or self.REDIS_SERVER.startswith("rediss://")
        ):
            return self.REDIS_SERVER
        return "redis://{}/".format(self.REDIS_SERVER)

    def __init__(self):
        if self.REDIS_SENTINEL:
            sentinels = [tuple(s.split(':')) for s in self.REDIS_SENTINEL.split(';')]
            self._sentinel = redis.sentinel.Sentinel(sentinels,
                                                     db=self.REDIS_SENTINEL_DB,
                                                     socket_timeout=0.1,
                                                     sentinel_kwargs={"password": self.REDIS_SENTINEL_PASSWORD}
                                                     )
        else:
            self._redis = redis.Redis.from_url(self.rewrite_redis_url())

    def get_async_redis_client(self):
        conn = self.get_redis().connection_pool.get_connection("ping")
        host, port, password = conn.host, conn.port, conn.password
        r = aioredis.Redis(host=host, port=port, password=password)
        return r.pubsub()

    def get_host_identifier(self):
        """\
        we try to return IPADDR:PID form to identify where any singlebeat instance is
        running.

        :return:
        """
        if self._host_identifier:
            return self._host_identifier
        local_ip_addr = (
            self.get_redis()
            .connection_pool.get_connection("ping")
            ._sock.getsockname()[0]
        )
        self._host_identifier = "{}:{}".format(local_ip_addr, os.getpid())
        return self._host_identifier


config = Config()
config.checks()

numeric_log_level = getattr(logging, config.LOG_LEVEL.upper(), None)
logging.basicConfig(level=numeric_log_level)
logger = logging.getLogger(__name__)


def get_process_identifier(args):
    """by looking at arguments we try to generate a proper identifier
    >>> get_process_identifier(['python', 'echo.py', '1'])
    'python_echo.py_1'
    """
    return "_".join(args)


class State:
    PAUSED = "PAUSED"
    RUNNING = "RUNNING"
    WAITING = "WAITING"
    RESTARTING = "RESTARTING"


def is_process_alive(pid):
    try:
        os.kill(pid, 0)
        return True
    except:
        return False


class Process(object):
    def __init__(self, args):
        self.args = args
        self.state = None
        self.t1 = time.time()

        self.identifier = config.IDENTIFIER or get_process_identifier(self.args[1:])
        self.ioloop = asyncio.get_running_loop()

        for signame in {"SIGINT", "SIGTERM"}:
            sig = getattr(signal, signame)
            self.ioloop.add_signal_handler(
                sig, functools.partial(self.sigterm_handler, sig, self.ioloop)
            )

        self.async_redis = config.get_async_redis_client()
        self.fence_token = 0
        self.sprocess = None
        self.pc = None
        self.state = State.WAITING
        self._periodic_callback_running = True
        self.child_exit_cb = self.proc_exit_cb

    def proc_exit_cb(self, exit_status):
        """When child exits we use the same exit status code"""
        self._periodic_callback_running = False
        sys.exit(exit_status)

    def proc_exit_cb_noop(self, exit_status):
        """\
        when we deliberately restart/stop the child process,
        we don't want to exit ourselves, so we replace proc_exit_cb
        with a noop one when restarting
        :param exit_status:
        :return:
        """

    def proc_exit_cb_restart(self, exit_status):
        """\
        this is used when we restart the process,
        it re-triggers the start
        """
        self.ioloop.run_until_complete(self.spawn_process())

    def proc_exit_cb_state_set(self, exit_status):
        if self.state == State.PAUSED:
            self.state = State.WAITING

    def stdout_read_cb(self, data):
        sys.stdout.write(data)

    def stderr_read_cb(self, data):
        sys.stderr.write(data)

    async def timer_cb_paused(self):
        pass

    async def timer_cb_waiting(self):
        if self.acquire_lock():
            logger.info(f"acquired lock, {self.identifier} spawning child process")
            return self.ioloop.create_task(self.spawn_process())
        # couldn't acquire lock
        if config.WAIT_MODE == "supervised":
            logger.debug(
                "already running, will exit after %s seconds" % config.WAIT_BEFORE_DIE
            )
            time.sleep(config.WAIT_BEFORE_DIE)
            sys.exit()

    def process_pid(self):
        """\
        when we are restarting, we want to keep sending heart beat, so any other single-beat
        node will not pick it up.
        hence we need a process-id as an identifier - even for a short period of time.
        :return:
        """
        if self.sprocess:
            return self.sprocess.pid
        return -1

    async def timer_cb_running(self):
        rds = config.get_redis()
        # read current fence token
        redis_fence_token = rds.get(
            "SINGLE_BEAT_{identifier}".format(identifier=self.identifier)
        )

        if redis_fence_token:
            redis_fence_token = int(redis_fence_token.split(b":")[0])
        else:
            logger.error(
                "fence token could not be read from Redis - assuming lock expired, trying to reacquire lock"
            )
            if self.acquire_lock():
                logger.info("reacquired lock")
                redis_fence_token = self.fence_token
            else:
                logger.error("unable to reacquire lock, terminating")
                os.kill(os.getpid(), signal.SIGTERM)

        logger.debug(
            "expected fence token: {} fence token read from Redis: {}".format(
                self.fence_token, redis_fence_token
            )
        )

        if self.fence_token == redis_fence_token:
            self.fence_token += 1
            rds.set(
                "SINGLE_BEAT_{identifier}".format(identifier=self.identifier),
                "{}:{}:{}".format(
                    self.fence_token, config.HOST_IDENTIFIER, self.process_pid()
                ),
                ex=config.LOCK_TIME,
            )
        else:
            logger.error(
                "fence token did not match - lock is held by another process, terminating"
            )
            # send sigterm to ourself and let the sigterm_handler do the rest
            os.kill(os.getpid(), signal.SIGTERM)

    async def timer_cb_restarting(self):
        """\
        when restarting we are doing exactly the same as running - we don't want any other
        single-beat node to pick up
        :return:
        """
        await self.timer_cb_running()

    async def timer_cb(self):
        logger.debug("timer called %s state=%s", time.time() - self.t1, self.state)
        self.t1 = time.time()
        fn = getattr(self, "timer_cb_{}".format(self.state.lower()))
        await fn()

    def acquire_lock(self):
        rds = config.get_redis()
        return rds.execute_command(
            "SET",
            "SINGLE_BEAT_{}".format(self.identifier),
            "{}:{}:{}".format(self.fence_token, config.HOST_IDENTIFIER, 0),
            "NX",
            "EX",
            config.INITIAL_LOCK_TIME,
        )

    def sigterm_handler(self, signum, loop):
        """When we get term signal
        if we are waiting and got a sigterm, we just exit.
        if we have a child running, we pass the signal first to the child
        then we exit.

        To exit we signal our main sleep/trigger loop on `self.run()`

        :param signum:
        :param ioloop:
        :return:
        """
        assert self.state in ("WAITING", "RUNNING", "PAUSED")
        logger.debug("our state %s", self.state)
        if self.state == "WAITING":
            self._periodic_callback_running = False

        if self.state == "RUNNING":
            logger.debug(
                "already running sending signal to child - %s", self.sprocess.pid
            )
            self.sprocess.send_signal(signum)
            logger.debug("waiting for subprocess to finish")
            self.ioloop.create_task(self.sprocess.wait())
        self._periodic_callback_running = False

    async def run(self):
        while self._periodic_callback_running:
            await self.timer_cb()
            await asyncio.sleep(config.HEARTBEAT_INTERVAL)

    async def _read_stream(self, stream, cb):
        decoder = codecs.getincrementaldecoder('utf-8')(errors='strict')

        while True:
            line = await stream.read(100)
            if line:
                cb(decoder.decode(line))
            else:
                break

    async def spawn_process(self):
        cmd = self.args
        env = os.environ

        self.state = State.RUNNING
        try:
            self.sprocess = await asyncio.create_subprocess_exec(
                *cmd,
                env=env,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
        except FileNotFoundError:
            """
            if the file that we need to run doesn't exists
            we immediately exit.
            """
            logger.exception("file not found")
            return self.child_exit_cb(1)
        try:
            await asyncio.wait(
                [
                    asyncio.create_task(self._read_stream(self.sprocess.stdout, self.forward_stdout)),
                    asyncio.create_task(self._read_stream(self.sprocess.stderr, self.forward_stderr)),
                ]
            )
            self.child_exit_cb(self.sprocess.returncode)
        except SystemExit as e:
            os._exit(e.code)

    def cli_command_info(self, msg):
        info = ""
        if self.sprocess:
            if is_process_alive(self.sprocess.pid):
                info = "pid: {}".format(self.sprocess.pid)
        return info

    def child_process_alive(self):
        return not self.sprocess.protocol._process_exited

    def cli_command_quit(self, msg):
        """\
        kills the child and exits
        """
        if self.state == State.RUNNING and self.sprocess and self.child_process_alive():
            self.sprocess.kill()
        else:
            sys.exit(0)

    def cli_command_pause(self, msg):
        """\
        if we have a running child we kill it and set our state to paused
        if we don't have a running child, we set our state to paused
        this will pause all the nodes in single-beat cluster

        its useful when you deploy some code and don't want your child to spawn
        randomly

        :param msg:
        :return:
        """
        info = ""
        if self.state == State.RUNNING and self.sprocess and self.child_process_alive():
            self.child_exit_cb = self.proc_exit_cb_noop
            self.sprocess.kill()
            info = "killed"
            # TODO: check if process is really dead etc.
        self.state = State.PAUSED
        return info

    def cli_command_resume(self, msg):
        """\
        sets state to waiting - so we resume spawning children
        """
        if self.state == State.PAUSED:
            self.state = State.WAITING

    def cli_command_stop(self, msg):
        """\
        stops the running child process - if its running
        it will re-spawn in any single-beat node after sometime

        :param msg:
        :return:
        """
        info = ""
        if self.state == State.RUNNING and self.sprocess and self.sprocess.proc:
            self.state = State.PAUSED
            # TODO:
            # self.sprocess.set_exit_callback(self.proc_exit_cb_state_set)
            self.sprocess.kill()
            info = "killed"
            # TODO: check if process is really dead etc.
        return info

    def cli_command_restart(self, msg):
        """\
        restart the subprocess
        i. we set our state to RESTARTING - on restarting we still send heartbeat
        ii. we kill the subprocess
        iii. we start again
        iv. if its started we set our state to RUNNING, else we set it to WAITING

        :param msg:
        :return:
        """
        info = ""
        if self.state == State.RUNNING and self.sprocess and self.sprocess.proc:
            self.state = State.RESTARTING
            self.child_exit_cb = self.proc_exit_cb_restart
            self.sprocess.kill()
            info = "killed"
            # TODO: check if process is really dead etc.
        return info

    def pubsub_callback(self, msg):
        logger.info("got command - %s", msg)

        if msg["type"] != b"message":
            return

        try:
            cmd = json.loads(msg["data"])
        except:
            logger.exception("exception on parsing command %s", msg)
            return

        fn = getattr(self, "cli_command_{}".format(cmd["cmd"]), None)
        if not fn:
            logger.info("cli_command_{} not found".format(cmd["cmd"]))
            return

        logger.info("got command - %s running %s", msg["data"], fn)
        info = fn(cmd)
        rds = config.get_redis()
        logger.info("reply to %s", cmd["reply_channel"])
        rds.publish(
            cmd["reply_channel"],
            json.dumps(
                {
                    "identifier": config.get_host_identifier(),
                    "state": self.state,
                    "info": info or "",
                }
            ),
        )

    async def wait_for_commands(self):
        logger.info("subscribed to %s", "SB_{}".format(self.identifier))
        await self.async_redis.subscribe("SB_{}".format(self.identifier))
        logger.debug("subscribed to redis channel %s", "SB_{}".format(self.identifier))
        async for msg in self.async_redis.listen():
            self.pubsub_callback(msg)

    def forward_stdout(self, buf):
        self.stdout_read_cb(buf)

    def forward_stderr(self, buf):
        self.stderr_read_cb(buf)


async def run_process():
    process = Process(sys.argv[1:])
    await process.run()


def main():
    asyncio.run(run_process())


if __name__ == "__main__":
    main()

